# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/MediaEngineWebRTCAudio.h
# Commit: bd01b1a3f7d7
# Full Hash: bd01b1a3f7d76f5d461cd55d97b097a1ead66206
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2020-11-06 04:14:34
# Regressor Bug: 1651745
# File Overlap Count: 1
# Description:
#   Bug 1651745 - Simplify the same-iteration assertions in AudioInputProcessing. r=padenot
#   
#   With a dedicated MediaTrack subclass for microphone input we can now coordinate
#   appending real data and pulling silence in a single place. This makes it easier
#   to control the amount of buffering needed, and the timing expectations around
# ==============================================================================

diff -r c504995adbb4 -r bd01b1a3f7d7 dom/media/webrtc/MediaEngineWebRTCAudio.h
--- a/dom/media/webrtc/MediaEngineWebRTCAudio.h	Thu Nov 05 16:42:28 2020 +0000
+++ b/dom/media/webrtc/MediaEngineWebRTCAudio.h	Thu Nov 05 16:59:51 2020 +0000
@@ -140,7 +140,8 @@
                        const PrincipalHandle& aPrincipalHandle);
 
   void Pull(MediaTrackGraphImpl* aGraph, GraphTime aFrom, GraphTime aTo,
-            GraphTime aTrackEnd, AudioSegment* aSegment, bool* aEnded);
+            GraphTime aTrackEnd, AudioSegment* aSegment,
+            bool aLastPullThisIteration, bool* aEnded);
 
   void NotifyOutputData(MediaTrackGraphImpl* aGraph, AudioDataValue* aBuffer,
                         size_t aFrames, TrackRate aRate,
@@ -228,17 +229,15 @@
   AlignedFloatBuffer mInputDownmixBuffer;
   // Stores data waiting to be pulled.
   AudioSegment mSegment;
-#ifdef DEBUG
-  // The MTGImpl::IterationEnd() of the last time we appended data from an
-  // audio callback.
-  GraphTime mLastCallbackAppendTime;
-#endif
   // Set to false by Start(). Becomes true after the first time we append real
   // audio frames from the audio callback.
   bool mLiveFramesAppended;
-  // Set to false by Start(). Becomes true after the first time we append
-  // silence *after* the first audio callback has appended real frames.
-  bool mLiveSilenceAppended;
+  // Once live frames have been appended, this is the number of frames appended
+  // as pre-buffer for that data, to avoid underruns. Buffering in the track
+  // might be needed because of the AUDIO_BLOCK interval at which we run the
+  // graph, the packetizer keeping some input data. Care must be taken when
+  // turning on and off the packetizer.
+  TrackTime mLiveBufferingAppended;
   // Principal for the data that flows through this class.
   const PrincipalHandle mPrincipal;
   // Whether or not this MediaEngine is enabled. If it's not enabled, it
